nft_ingester/src/backfiller.rs

```diff
@@ -0,0 +1,702 @@
+struct BackfillTree {
+    unique_tree: UniqueTree,
+    backfill_from_seq_1: bool,
+}
+
+impl BackfillTree {
+    fn new(unique_tree: UniqueTree, backfill_from_seq_1: bool) -> Self {
+        Self {
+            unique_tree,
+            backfill_from_seq_1,
+        }
+    }
+}
+
+/// Struct used when querying the max sequence number of a tree.
+#[derive(Debug, FromQueryResult, Clone)]
+struct MaxSeqItem {
+    max: i64,
+}
+
+/// Struct used when querying for items to backfill.
+#[derive(Debug, FromQueryResult, Clone)]
+struct SimpleBackfillItem {
+    seq: i64,
+    slot: i64,
+}
+
+/// Struct used to store sequence number gap info for a given tree.
+struct GapInfo {
+    prev: SimpleBackfillItem,
+    curr: SimpleBackfillItem,
+}
+
+impl GapInfo {
+    fn new(prev: SimpleBackfillItem, curr: SimpleBackfillItem) -> Self {
+        Self { prev, curr }
+    }
+}
+
+/// Main struct used for backfiller task.
+struct Backfiller<T: Messenger> {
+    db: DatabaseConnection,
+    listener: PgListener,
+    rpc_client: RpcClient,
+    messenger: T,
+    failure_delay: u64,
+}
+
+impl<T: Messenger> Backfiller<T> {
+    /// Create a new `Backfiller` struct.
+    async fn new(pool: Pool<Postgres>, config: IngesterConfig) -> Self {
+        // Create Sea ORM database connection used later for queries.
+        let db = SqlxPostgresConnector::from_sqlx_postgres_pool(pool.clone());
+
+        // Connect to database using sqlx and create PgListener.
+        let mut listener = sqlx::postgres::PgListener::connect_with(&pool.clone())
+            .await
+            .map_err(|e| IngesterError::StorageListenerError {
+                msg: format!("Could not connect to db for PgListener {e}"),
+            })
+            .unwrap();
+
+        // Get database listener channel.
+        let channel = config
+            .database_config
+            .get(&*DATABASE_LISTENER_CHANNEL_KEY)
+            .and_then(|u| u.clone().into_string())
+            .ok_or(IngesterError::ConfigurationError {
+                msg: format!(
+                    "Database listener channel missing: {}",
+                    DATABASE_LISTENER_CHANNEL_KEY
+                ),
+            })
+            .unwrap();
+
+        // Setup listener on channel.
+        listener
+            .listen(&channel)
+            .await
+            .map_err(|e| IngesterError::StorageListenerError {
+                msg: format!("Error listening to channel on backfill_items tbl {e}"),
+            })
+            .unwrap();
+
+        // Get RPC URL.
+        let rpc_url = config
+            .rpc_config
+            .get(&*RPC_URL_KEY)
+            .and_then(|u| u.clone().into_string())
+            .ok_or(IngesterError::ConfigurationError {
+                msg: format!("RPC URL missing: {}", RPC_URL_KEY),
+            })
+            .unwrap();
+
+        // Instantiate RPC client.
+        let rpc_client = RpcClient::new_with_commitment(rpc_url, CommitmentConfig::processed());
```

Comment:
Yes, so far I have found that actually when I call `get_block()` its assuming finalized even now when this is set to processed, even though this client is set to processed.

I was experimenting on why `get_block()` was doing that and accidentally left this at processed here.  Good catch.

Reply:
Yes I meant to have this at a higher level, but was experimenting and forgot to change it back.  Good catch.

The reason I was experimenting is because I found that even if it is set to `processed` or `confirmed` here, my call to `get_block()` defaults to `finalized`.  I think I would need to call one of the functions like `get_block_with_config()` and set it for that call.